drivers/watchdog/sunxi-twd.c

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright © 2017-2018 The Crust Firmware Authors.
+ * SPDX-License-Identifier: (BSD-3-Clause OR GPL-2.0)
+ */
+
+#include <error.h>
+#include <mmio.h>
+#include <drivers/clock.h>
+#include <drivers/timer.h>
+#include <drivers/watchdog.h>
+#include <drivers/watchdog/sunxi-twd.h>
+
+#define CLOCK_CYCLES_MHZ    24
+
+#define TWD_STATUS_REG      0x00
+#define TWD_CTRL_REG        0x10
+#define TWD_RESTART_REG     0x14
+#define TWD_INTV_VAL_REG    0x30
+
+#define TWD_RESTART_KEY_VAL (0xD14 << 16)
+
+static void
+sunxi_twd_restart(void *param)
+{
+	uint32_t reg;
+
+	/* Cast param from periodic work function interface. */
+	struct device *dev = (struct device *)param;
+
+	/* Enable and perform restart. */
+	reg = TWD_RESTART_KEY_VAL | BIT(0);
+	mmio_write32(dev->regs + TWD_RESTART_REG, reg);
+}
+
+static int
+sunxi_twd_disable(struct device *dev)
+{
+	/* Cancel TWD restart timer. */
+	timer_cancel_periodic(sunxi_twd_restart, &dev);
```

Comment:
You should comment why you're intentionally ignoring the return value of `timer_cancel_periodic` (something like because `sunxi_twd_disable` is called from the probe function when the periodic work is not registered).

Also, this should definitely happen *after* stopping the watchdog.